controllers/internal/utils/dns.go

```diff
@@ -20,30 +20,99 @@ package utils
 import (
 	"fmt"
 	"sort"
+	"strings"
 
 	"github.com/lixiangzhong/dnsutil"
+	"github.com/miekg/dns"
 )
 
+type DNSServer struct {
+	Host string
+	Port int
+}
+
 // Dig retrieves list of tuple <IP address, A record > from edge DNS server for specific FQDN
 func Dig(edgeDNSServer, fqdn string) ([]string, error) {
-	var dig dnsutil.Dig
 	if edgeDNSServer == "" {
 		return nil, fmt.Errorf("empty edgeDNSServer")
 	}
-	err := dig.SetDNS(edgeDNSServer)
-	if err != nil {
-		err = fmt.Errorf("dig error: can't set query dns (%s) with error(%s)", edgeDNSServer, err)
-		return nil, err
+	dnsServer := DNSServer{
+		Host: edgeDNSServer,
+		Port: 53,
+	}
+	return DigWithFallback([]DNSServer{dnsServer}, fqdn)
+}
+
+func DigWithFallback(edgeDNSServers []DNSServer, fqdn string) ([]string, error) {
+	var dig dnsutil.Dig
+	if len(edgeDNSServers) == 0 {
+		return nil, fmt.Errorf("empty edgeDNSServers, provide at least one")
 	}
-	a, err := dig.A(fqdn)
-	if err != nil {
-		err = fmt.Errorf("dig error: can't dig fqdn(%s) with error(%s)", fqdn, err)
-		return nil, err
+	for i := 0; i < len(edgeDNSServers); i += 2 {
+		if edgeDNSServers[i].Host == "" {
+			continue
+		}
+		err := dig.SetDNS(OneToString(edgeDNSServers[i]))
+		if err != nil {
+			err = fmt.Errorf("dig error: can't set query dns (%s) with error(%s)",
+				OneToString(edgeDNSServers[i]), err)
+			return nil, err
+		}
+		if len(edgeDNSServers) > i+1 {
+			backupDNSServer := edgeDNSServers[i+1]
+			if backupDNSServer.Host == "" {
+				return nil, fmt.Errorf("empty edgeDNSServer")
+			}
+			err := dig.SetBackupDNS(OneToString(backupDNSServer))
+			if err != nil {
+				err = fmt.Errorf("dig error: can't set dns backup server (%s) with error(%s)",
+					OneToString(backupDNSServer), err)
+				return nil, err
+			}
+		}
+		a, err := dig.A(fqdn)
+		if err != nil {
+			if len(edgeDNSServers) > i+2 { // there are still fallback dns servers listed
+				continue
+			} else {
+				err = fmt.Errorf("dig error: can't dig fqdn(%s) with error(%s)", fqdn, err)
+				return nil, err
+			}
+		}
+		var IPs []string
```

Comment:
`var ips []string` local variable

Reply:
couldn't this be catched actually by some of our linters?